internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,455 @@
+	CPUWeightMin = 1
+	CPUWeightMax = 9
+)
+
+var (
+	ErrAlreadyClosed = errors.New("the handle has already been closed")
+	ErrNotRegistered = errors.New("job is not registered to receive notifications")
+)
+
+// Create creates a job object.
+//
+// `name` specifies the name of the job object if a named job object is desired. If name
+// is an empty string, the job will not be assigned a name.
+//
+// `notifications` specifies if the job will be registered to receive notifications.
+// If this is false, `PollNotifications` will return immediately with error `errNotRegistered`.
+//
+// Returns a JobObject structure and an error if there is one.
+func Create(ctx context.Context, name string, notifications bool) (_ *JobObject, err error) {
+	// If the IOCP we'll be using to receive messages for all jobs hasn't been
+	// created, create it and start polling.
+	ioInitOnce.Do(func() {
+		h, err := windows.CreateIoCompletionPort(windows.InvalidHandle, 0, 0, 0xffffffff)
+		if err != nil {
+			initIOErr = err
+			return
+		}
+		ioCompletionPort = h
+		go pollIOCP(ctx, h)
+	})
+
+	if initIOErr != nil && notifications {
+		return nil, initIOErr
+	}
+
+	var jobName *uint16
+	if name != "" {
+		jobName, err = windows.UTF16PtrFromString(name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	jobHandle, err := windows.CreateJobObject(nil, jobName)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			windows.Close(jobHandle)
+			delete(jobMap, uintptr(jobHandle))
+		}
+	}()
+
+	var mq *queue.MessageQueue
+	if notifications {
+		mq = queue.NewMessageQueue()
+		jobMap[uintptr(jobHandle)] = mq
+		if err = attachIOCP(jobHandle, ioCompletionPort); err != nil {
+			return nil, err
+		}
+	}
+
+	return &JobObject{
+		handle: jobHandle,
+		mq:     mq,
+	}, nil
+}
+
+// SetResourceLimits sets resource limits on the job object (cpu, memory, storage).
+func (job *JobObject) SetResourceLimits(limits *JobLimits) error {
+	// Go through and check what limits were specified and apply them to the job.
+	if limits.MemoryLimitInBytes != 0 {
+		if err := job.SetMemoryLimit(limits.MemoryLimitInBytes); err != nil {
+			return errors.Wrap(err, "failed to set job object memory limit")
+		}
+	}
+
+	if limits.CPULimit != 0 {
+		if err := job.SetCPULimit(RateBased, limits.CPULimit); err != nil {
+			return errors.Wrap(err, "failed to set job object cpu limit")
+		}
+	} else if limits.CPUWeight != 0 {
+		if err := job.SetCPULimit(WeightBased, limits.CPUWeight); err != nil {
+			return errors.Wrap(err, "failed to set job object cpu limit")
+		}
+	}
+
+	if limits.MaxBandwidth != 0 || limits.MaxIOPS != 0 {
+		if err := job.SetIOLimit(limits.MaxBandwidth, limits.MaxIOPS); err != nil {
+			return errors.Wrap(err, "failed to set io limit on job object")
+		}
+	}
+	return nil
+}
+
+// SetCPULimit sets the CPU limit specified on the job object.
+func (job *JobObject) SetCPULimit(rateControlType CPURateControlType, rateControlValue uint32) error {
```

Comment:
I feel like using this `CPURateControlType` is possibly a bit unnecessary? There's (probably) never going to be a type internally other than WeightBased or RateBased, so created a "CpuRateControlType" doesn't feel like it helps that much. Could we not instead pass in the `winapi.JOB_OBJECT_CPU_RATE_CONTROL_*` value for the corresponding control type and use a switch statement in here?

Reply:
The reason for the type is so that they can only pass in these two values and nothing else. The JOB_OBJECT_CPU_RATE_CONTROL_* flags are just ints so you could pass in anything for these, and it's not really clear to the caller what to pass in here if they're unfamiliar with the API and I don't feel like a comment explaining to pass in one of these is better than putting a type for the parameter . Just locking it so that you only have two options, being the only two options their are, seems ok to me.